### PR TITLE
Fix #5795: Implement sprite cycle checking v2

### LINKED ISSUE #5795: The game freezes with maptile (0,0) on screen after an autosave
**OS:** [Windows 10]
**Version:** [0.0.8]
**Commit/Build:** [4e4022d693d4149eddea906ec71f22ddc626411f]

If the appended scenario is loaded, an autosave has occurred and maptile (0,0) is in the active viewport, the game freezes

**Steps to reproduce:**
1. load the appended scenario.
2. wait for an autosave.
3. pan the camera to the top of the map.

**Save game:**
[Sandbox.zip](https://github.com/OpenRCT2/OpenRCT2/files/1123149/Sandbox.zip)

### PR DESCRIPTION
**Pull request by janisozaur**
This is #5801 rebased and awaiting improvements as mentioned here: https://github.com/OpenRCT2/OpenRCT2/pull/5801#issuecomment-313341587

Needs network version, which was not changed so far.

### CODE REVIEW

File: src/openrct2/rct2/S6Exporter.cpp [hunk 1]
```diff
@@ -152,6 +152,7 @@ void S6Exporter::Save(IStream * stream, bool isScenario)
 
 void S6Exporter::Export()
 {
+    openrct2_assert(!(check_for_spatial_index_cycles(false) || check_for_sprite_list_cycles(false)), "A sprite cycle exists.");
```
Comment (IntelOrca) [inline src/openrct2/rct2/S6Exporter.cpp]: Could it say where the cycle exists? E.g. `A sprite cycle exists in the spatial index`

File: src/openrct2/rct2/S6Exporter.cpp [hunk 2]
```diff
@@ -449,7 +450,6 @@ extern "C"
 
         map_reorganise_elements();
         sprite_clear_all_unused();
```
Comment (IntelOrca) [inline src/openrct2/rct2/S6Exporter.cpp]: Remove this

File: src/openrct2/world/sprite.c [hunk 3]
```diff
@@ -285,8 +261,39 @@ static void sprite_reset(rct_unk_sprite *sprite)
...
     sprite->previous = prev;
     sprite->sprite_index = sprite_index;
+    sprite->sprite_identifier = SPRITE_IDENTIFIER_NULL;
+}
+
+/**
+* Clears all the unused sprite memory to zero. Probably so that it can be compressed better when saving.
+*  rct2: 0x0069EBA4
+*/
+void sprite_clear_all_unused()
```
Comment (IntelOrca) [inline src/openrct2/world/sprite.c]: Merge this with `reset_empty_sprites`

Comment (janisozaur) [inline src/openrct2/world/sprite.c]: The two are very similar, but `sprite_clear_all_unused` is the more conservative version: it explicitly sets `linked_list_type_offset`, `sprite_index` and `next_in_quadrant` (if broken) fields. I don't want to change this behaviour this close to release, are you fine with `sprite_clear_all_unused` calling `reset_empty_sprites`, but then still doing a post-process loop?

Comment (IntelOrca) [inline src/openrct2/world/sprite.c]: If you think its too risky, then remove the reset version for this release and then merge them for next release. I don't see the need for both.

**Comment by rwjuk** [comment]
@janisozaur I'm not entirely sure how I'd go about implementing the reordering of the list, so if you have a plan for it in mind, go for it.

**Comment by janisozaur** [comment]
I've implemented some fixes, there's still an issue that I have not solved.

Since the previous code could reset any sprite, it may have reset sprite at some index `x != 0`, which would set its `sprite_index` to 0, I have checked and looks like all places expect this be the valid number. Such sprites would also not be a part of any regular (non-spatial) list, because if there was any link to them, it got invalidated. This needs to be taken care of too.

File: src/openrct2/world/sprite.c (hunk 3 above)
Comment (janisozaur) [inline src/openrct2/world/sprite.c]: I removed `reset_empty_sprites` after inspecting other sprite handling code.

**Comment by janisozaur** [comment]
That should hopefully be all there is. I tested it a little with some of the parks I have and didn't see anything suspicious, but please @rwjuk @PFCKrutonium @IntelOrca do test it!

File: src/openrct2/world/sprite.c [hunk 4]
```diff
@@ -951,6 +951,50 @@ sint32 check_for_sprite_list_cycles(bool fix)
...
 
+/**
+ * Finds and fixes null sprites that are not reachable via SPRITE_LIST_NULL list.
+ *
+ * @return count of disjoint sprites found
+ */
+sint32 fix_disjoint_sprites()
+{
+    // Find reachable sprites
+    bool reachable[MAX_SPRITES] = {};
```
Comment (rwjuk) [inline src/openrct2/world/sprite.c]: `{ 0 }`

Comment (janisozaur) [inline src/openrct2/world/sprite.c]: Why? `{}` already includes initial element.

Comment (rwjuk) [inline src/openrct2/world/sprite.c]: Good question, but it refuses to build with MSVC (see Appveyor) without it. It's stupid, but then MSVC often is.

Comment (IntelOrca) [inline src/openrct2/world/sprite.c]: I think its because the CIs still use VS 2015.

Comment (rwjuk) [inline src/openrct2/world/sprite.c]: @IntelOrca My VS 2017 doesn't like it either.

Comment (IntelOrca) [inline src/openrct2/world/sprite.c]: Oh, I tried it using https://gcc.godbolt.org with CL compiler and it worked.

**Comment by rwjuk** [comment]
Have tested this with a bunch of SV6 and SC6, nothing untoward has happened so far.

**Pull request merged**